patches/server/1048-add-ItemFrameMeta-API.patch

```diff
@@ -0,0 +1,323 @@
+         }
++        if (itemType == ItemType.ITEM_FRAME || itemType == ItemType.GLOW_ITEM_FRAME) {
++            return CraftItemMetas.asType(CraftItemMetas.ITEM_FRAME_META_DATA);
++        }
++        // Paper end - add ItemFrameMeta API
+         if (itemType == ItemType.COMPASS) {
+             return CraftItemMetas.asType(CraftItemMetas.COMPASS_META_DATA);
+         }
+diff --git a/src/main/java/org/bukkit/craftbukkit/inventory/CraftMetaEntityTag.java b/src/main/java/org/bukkit/craftbukkit/inventory/CraftMetaEntityTag.java
+index da474a5b963d8e6769d120e9091e60ed0a468a9f..c113001a2353ec1234a18c3291504f48389482be 100644
+--- a/src/main/java/org/bukkit/craftbukkit/inventory/CraftMetaEntityTag.java
++++ b/src/main/java/org/bukkit/craftbukkit/inventory/CraftMetaEntityTag.java
+@@ -20,8 +20,10 @@ public class CraftMetaEntityTag extends CraftMetaItem {
+             Material.PUFFERFISH_BUCKET,
+             Material.SALMON_BUCKET,
+             Material.TADPOLE_BUCKET, // Paper
+-            Material.ITEM_FRAME,
+-            Material.GLOW_ITEM_FRAME,
++            // Paper start - add ItemFrameMeta API
++            // Material.ITEM_FRAME,
++            // Material.GLOW_ITEM_FRAME,
++            // Paper end - add ItemFrameMeta API
+             Material.PAINTING
+     );
+ 
+diff --git a/src/main/java/org/bukkit/craftbukkit/inventory/CraftMetaItem.java b/src/main/java/org/bukkit/craftbukkit/inventory/CraftMetaItem.java
+index 5c45946f54e4b16440c15a9165ef85be43e53c84..07ad086bfa77a4b3bcaa188ea65ceb7250da9c04 100644
+--- a/src/main/java/org/bukkit/craftbukkit/inventory/CraftMetaItem.java
++++ b/src/main/java/org/bukkit/craftbukkit/inventory/CraftMetaItem.java
+@@ -2132,6 +2132,7 @@ class CraftMetaItem implements ItemMeta, Damageable, Repairable, BlockDataMeta {
+                 map.put(CraftMetaEnchantedBook.class, Set.of(CraftMetaEnchantedBook.STORED_ENCHANTMENTS.TYPE));
+                 map.put(CraftMetaEntityTag.class, Set.of(CraftMetaEntityTag.ENTITY_TAG.TYPE));
+                 map.put(CraftMetaFirework.class, Set.of(CraftMetaFirework.FIREWORKS.TYPE));
++                map.put(CraftMetaItemFrame.class, Set.of(CraftMetaItemFrame.ENTITY_TAG.TYPE)); // Paper - add ItemFrameMeta API
+                 map.put(CraftMetaKnowledgeBook.class, Set.of(CraftMetaKnowledgeBook.BOOK_RECIPES.TYPE));
+                 map.put(CraftMetaLeatherArmor.class, Set.of(CraftMetaLeatherArmor.COLOR.TYPE));
+                 map.put(CraftMetaMap.class, Set.of(CraftMetaMap.MAP_COLOR.TYPE, CraftMetaMap.MAP_POST_PROCESSING.TYPE, CraftMetaMap.MAP_ID.TYPE));
+diff --git a/src/main/java/org/bukkit/craftbukkit/inventory/CraftMetaItemFrame.java b/src/main/java/org/bukkit/craftbukkit/inventory/CraftMetaItemFrame.java
+new file mode 100644
+index 0000000000000000000000000000000000000000..b13444c63c37e5c9f4ff969ad473f9280a961332
+--- /dev/null
++++ b/src/main/java/org/bukkit/craftbukkit/inventory/CraftMetaItemFrame.java
```

Comment:
Isn't it better to put this into some paper package since this is not a CraftBukkit addition?

Reply:
I mentioned up above in my original PR. The issue is that putting it in a different package requires stupid amounts of access transformers on CraftMetaItem and making nearly everything in CraftMetaItemFrame itself public.

Reply:
Well, I personally would probably put it into a paper package but I would also say that input by developers is needed here.

Reply:
Its fine in the CB package.